pulsar-client/src/main/java/org/apache/pulsar/client/impl/AutoClusterFailover.java

```diff
@@ -0,0 +1,234 @@
+    private long primaryRecoverTimestamp;
+    private long secondaryFailedTimestamp;
+    private final int timeout = 30_000;
+
+    private AutoClusterFailover(String primary, String secondary, long failoverDelayMs, long switchBackDelayMs) {
+        this.primary = primary;
+        this.secondary = secondary;
+        this.failoverDelayMs = failoverDelayMs;
+        this.switchBackDelayMs = switchBackDelayMs;
+        this.currentPulsarServiceUrl = primary;
+        this.primaryFailedTimestamp = -1;
+        this.primaryRecoverTimestamp = -1;
+        this.secondaryFailedTimestamp = -1;
+        this.timer = new Timer("pulsar-service-provider");
+    }
+
+    @Override
+    public void initialize(PulsarClient client) {
+        this.pulsarClient = client;
+
+        // start to probe primary cluster active or not
+        this.timer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                // current pulsar serviceUrl is primary
+                if (currentPulsarServiceUrl.equals(primary)) {
+                    if (probeAvailable(primary, timeout)) {
+                        primaryFailedTimestamp = -1;
+                        return;
+                    }
+
+                    if (primaryFailedTimestamp == -1) {
+                        primaryFailedTimestamp = System.currentTimeMillis();
+                    } else if (System.currentTimeMillis() - primaryFailedTimestamp < failoverDelayMs) {
+                        return;
+                    } else if (probeAvailable(secondary, timeout)){
+                        log.info("Current Pulsar service is primary: {}, it has been down for {} ms, "
+                                        + "switch to the secondary service: {}. The first primary service down at: {}",
+                                currentPulsarServiceUrl, System.currentTimeMillis() - primaryFailedTimestamp,
+                                secondary, primaryFailedTimestamp);
+                        try {
+                            pulsarClient.updateServiceUrl(secondary);
+                            currentPulsarServiceUrl = secondary;
+                        } catch (PulsarClientException e) {
+                            log.error("Failed to switch to secondary service URL ", e);
+                        }
+                    } else {
+                        log.error("Current Pulsar service is primary: {}, it has been down for {} ms. "
+                                + "Failed to switch to secondary service URL, "
+                                + "because secondary service URL is not available",
+                                currentPulsarServiceUrl, System.currentTimeMillis() - primaryFailedTimestamp);
+                    }
+                } else { // current pulsar service URL is secondary, probe whether we need to switch back to primary.
+                    if (!probeAvailable(currentPulsarServiceUrl, timeout)) {
```

Comment:
Can we make the url check and switch as a method? I think most of the following logic is duplicated with the primary service url check?

Reply:
done